dask/dataframe/shuffle.py

```diff
@@ -376,52 +388,136 @@ def set_partition(
     return df4.map_partitions(M.sort_index)
 
 
-def shuffle(
+class ShuffleRegistry:
+    """A registry of methods for shuffling a dataframe."""
+
+    class Method(NamedTuple):
+        can_use: Callable[[Optional[str], Any, List[str]], bool]
+        method: Callable
+        priority: int
+
+    def __init__(self):
+        self.methods: Set[ShuffleRegistry.Method] = set()
+
+    @property
+    def known_priorities(self):
+        return {m.priority for m in self.methods}
+
+    def register(
+        self,
+        priority: int,
+        can_use: Callable[[Optional[str], Any, List[str]], bool],
+        method: Callable,
+    ):
+        """Register a new shuffle method
+
+        Parameters
+        ----------
+        priority
+            The priority of this method, higher priority methods are tried first
+        can_use
+            Function to check if this method is useable for the given
+            shuffle parameters. Receives the requested shuffle name
+            (e.g. "tasks", "disk"), the index to shuffle on, and a
+            list of dataframe column names. Should return True if this
+            method can shuffle the dataframe, and False otherwise.
```

Comment:
I don't think I quite understand why the `priority` and `can_use` functionality is necessary here. At the moment, I get the impression that we can get away fine without it (but I could certainly be wrong).

Reply:
This was to be able to replicate the current functionality, but I suppose if we say "shuffle names are distinct", and you ask for them by name, then priority/can_use is unnecessary.